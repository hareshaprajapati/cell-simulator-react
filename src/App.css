.app {
  display: flex;
  flex-direction: column;
  align-items: center;
  background: black;
  height: 100vh;
  justify-content: center;
}

.grid:hover {
  cursor: pointer;
}

.row {
  display: flex;
  flex-direction: row;
}
.col {
  display: flex;
  flex-direction: column;
  background: #dee6e9;
  border: 2px solid #9e9ed4;
  padding: 16px;
}

.selected-column {
  background: blueviolet;
}

.not-selected-column {
  background: #dee6e9;
}

.next-generation-btn {
  background: #1976d2;
}

.btn {
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  color: white;
  padding: 6px 16px;
  border: 2px white solid;
}

.reset-btn {
  background: #757575;
  margin-left: 12px;
}

.btn-container {
  display: flex;
  margin: 8px;
}

.disable-btn {
  cursor: not-allowed;
}
.active-btn {
  cursor: pointer;
}
