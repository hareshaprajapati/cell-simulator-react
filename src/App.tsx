import React, { useState } from 'react';
import './App.css';
import { GRID_INIT_STATE } from './Constant';
import { arrayDeepCopy, countNeighbours } from './Utils';

const App = () => {

  const [grid, setGrid] = useState<number[][]>(arrayDeepCopy(GRID_INIT_STATE));

  // when user click on the div set grid's value 1 or 0 depends on current value
  const setConfig = (rowIndex: number, columnIndex: number) => {
    const gridCopy = arrayDeepCopy(grid);
    gridCopy[rowIndex][columnIndex] = gridCopy[rowIndex][columnIndex] === 0 ? 1 : 0;
    setGrid(gridCopy);
  }

  const resetConfig = () => {
    setGrid(arrayDeepCopy(GRID_INIT_STATE));
  }

  const generateNext = () => {
    const gridCopy = arrayDeepCopy(grid);

    for (let i = 0; i < grid.length; i++) {
      for (let j = 0; j < grid[0].length; j++) {
        const neighboursCount = countNeighbours(i, j, grid);
        if (gridCopy[i][j] === 1) {
          // A Cell with fewer than two live neighbours dies of under-population.
          if (neighboursCount < 2) {
            gridCopy[i][j] = 0;
          }
          // A Cell with 2 or 3 live neighbours lives on to the next generation. 
          if (neighboursCount === 2 || neighboursCount === 3) {
            gridCopy[i][j] = 1;
          }
          // A Cell with more than 3 live neighbours dies of overcrowding.
          if (neighboursCount > 3) {
            gridCopy[i][j] = 0;
          }
        }
        // An empty Cell with exactly 3 live neighbours comes to life.
        if (gridCopy[i][j] === 0 && neighboursCount === 3) {
          gridCopy[i][j] = 1;
        }
      }
    }
    setGrid(gridCopy);
  }

  const canGenerateNext = grid.some(row => row.includes(1));

  return (
    <div className="app">
      <div className="grid">
        {
          grid.map((row, rowIndex) => {
            return <div className="row" key={rowIndex}>
              {row.map((column, columnIndex) =>
                <div data-testid={`cell-${rowIndex}-${columnIndex}`} className={`col ${grid[rowIndex][columnIndex] === 1 ? 'selected-column' : ''}`} onClick={() => setConfig(rowIndex, columnIndex)}
                  key={`${rowIndex}-${columnIndex}`} />
              )}
            </div>
          })
        }
      </div>
      <div className="btn-container">
        <button data-testid="nextGenBtn" disabled={!canGenerateNext} className={`btn next-generation-btn ${!canGenerateNext ? 'disable-btn' : 'active-btn'} `} onClick={generateNext}>Next Generation</button>
        <button data-testid="resetBtn" disabled={!canGenerateNext} className={`btn reset-btn ${!canGenerateNext ? 'disable-btn' : 'active-btn'} `} onClick={resetConfig}>Reset</button>
      </div>
    </div >
  );
}

export default App;


