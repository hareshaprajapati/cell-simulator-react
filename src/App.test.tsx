import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import App from './App';

test('should validate cell classes', () => {
  const { container } = render(<App />);
  expect(container).toMatchSnapshot();
  const cell = screen.getByTestId('cell-0-0');
  expect(cell).toHaveClass('col');
  fireEvent.click(cell);
  expect(cell).toHaveClass('col selected-column');
});

test('should validate buttons', () => {
  render(<App />);
  const nextGenBtn = screen.getByTestId('nextGenBtn');
  const resetBtn = screen.getByTestId('resetBtn');
  // bydefault both the buttons should be disabled
  expect(nextGenBtn).toBeDisabled();
  expect(resetBtn).toBeDisabled();
  const cell = screen.getByTestId('cell-0-0');
  // after clicking on the cell both the buttons should be enabled
  fireEvent.click(cell);
  expect(nextGenBtn).toBeEnabled();
  expect(resetBtn).toBeEnabled();
  // by clicking on the reset button both the button should be disabled and cell have only col class 
  fireEvent.click(resetBtn);
  expect(cell).toHaveClass('col');
  expect(nextGenBtn).toBeDisabled();
  expect(resetBtn).toBeDisabled();
});
