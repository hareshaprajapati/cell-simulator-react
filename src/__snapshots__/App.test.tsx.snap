// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should validate cell class 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="grid"
    >
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-0-0"
        />
        <div
          class="col "
          data-testid="cell-0-1"
        />
        <div
          class="col "
          data-testid="cell-0-2"
        />
        <div
          class="col "
          data-testid="cell-0-3"
        />
        <div
          class="col "
          data-testid="cell-0-4"
        />
        <div
          class="col "
          data-testid="cell-0-5"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-1-0"
        />
        <div
          class="col "
          data-testid="cell-1-1"
        />
        <div
          class="col "
          data-testid="cell-1-2"
        />
        <div
          class="col "
          data-testid="cell-1-3"
        />
        <div
          class="col "
          data-testid="cell-1-4"
        />
        <div
          class="col "
          data-testid="cell-1-5"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-2-0"
        />
        <div
          class="col "
          data-testid="cell-2-1"
        />
        <div
          class="col "
          data-testid="cell-2-2"
        />
        <div
          class="col "
          data-testid="cell-2-3"
        />
        <div
          class="col "
          data-testid="cell-2-4"
        />
        <div
          class="col "
          data-testid="cell-2-5"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-3-0"
        />
        <div
          class="col "
          data-testid="cell-3-1"
        />
        <div
          class="col "
          data-testid="cell-3-2"
        />
        <div
          class="col "
          data-testid="cell-3-3"
        />
        <div
          class="col "
          data-testid="cell-3-4"
        />
        <div
          class="col "
          data-testid="cell-3-5"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-4-0"
        />
        <div
          class="col "
          data-testid="cell-4-1"
        />
        <div
          class="col "
          data-testid="cell-4-2"
        />
        <div
          class="col "
          data-testid="cell-4-3"
        />
        <div
          class="col "
          data-testid="cell-4-4"
        />
        <div
          class="col "
          data-testid="cell-4-5"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-5-0"
        />
        <div
          class="col "
          data-testid="cell-5-1"
        />
        <div
          class="col "
          data-testid="cell-5-2"
        />
        <div
          class="col "
          data-testid="cell-5-3"
        />
        <div
          class="col "
          data-testid="cell-5-4"
        />
        <div
          class="col "
          data-testid="cell-5-5"
        />
      </div>
    </div>
    <div
      class="btn-container"
    >
      <button
        class="btn next-generation-btn disable-btn "
        data-testid="nextGenBtn"
        disabled=""
      >
        Next Generation
      </button>
      <button
        class="btn reset-btn disable-btn "
        data-testid="resetBtn"
        disabled=""
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`should validate cell classes 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="grid"
    >
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-0-0"
        />
        <div
          class="col "
          data-testid="cell-0-1"
        />
        <div
          class="col "
          data-testid="cell-0-2"
        />
        <div
          class="col "
          data-testid="cell-0-3"
        />
        <div
          class="col "
          data-testid="cell-0-4"
        />
        <div
          class="col "
          data-testid="cell-0-5"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-1-0"
        />
        <div
          class="col "
          data-testid="cell-1-1"
        />
        <div
          class="col "
          data-testid="cell-1-2"
        />
        <div
          class="col "
          data-testid="cell-1-3"
        />
        <div
          class="col "
          data-testid="cell-1-4"
        />
        <div
          class="col "
          data-testid="cell-1-5"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-2-0"
        />
        <div
          class="col "
          data-testid="cell-2-1"
        />
        <div
          class="col "
          data-testid="cell-2-2"
        />
        <div
          class="col "
          data-testid="cell-2-3"
        />
        <div
          class="col "
          data-testid="cell-2-4"
        />
        <div
          class="col "
          data-testid="cell-2-5"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-3-0"
        />
        <div
          class="col "
          data-testid="cell-3-1"
        />
        <div
          class="col "
          data-testid="cell-3-2"
        />
        <div
          class="col "
          data-testid="cell-3-3"
        />
        <div
          class="col "
          data-testid="cell-3-4"
        />
        <div
          class="col "
          data-testid="cell-3-5"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-4-0"
        />
        <div
          class="col "
          data-testid="cell-4-1"
        />
        <div
          class="col "
          data-testid="cell-4-2"
        />
        <div
          class="col "
          data-testid="cell-4-3"
        />
        <div
          class="col "
          data-testid="cell-4-4"
        />
        <div
          class="col "
          data-testid="cell-4-5"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col "
          data-testid="cell-5-0"
        />
        <div
          class="col "
          data-testid="cell-5-1"
        />
        <div
          class="col "
          data-testid="cell-5-2"
        />
        <div
          class="col "
          data-testid="cell-5-3"
        />
        <div
          class="col "
          data-testid="cell-5-4"
        />
        <div
          class="col "
          data-testid="cell-5-5"
        />
      </div>
    </div>
    <div
      class="btn-container"
    >
      <button
        class="btn next-generation-btn disable-btn "
        data-testid="nextGenBtn"
        disabled=""
      >
        Next Generation
      </button>
      <button
        class="btn reset-btn disable-btn "
        data-testid="resetBtn"
        disabled=""
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
